### ISSUE TITLE
add svm in last layer

### ISSUE THREAD

**Issue by mundher** [fchollet/keras#2588]
I want to add svm in last layer of my model

```
model = Sequential()
model.add(Convolution2D(nb_filters, nb_conv, nb_conv,
border_mode='valid',
input_shape=(1, img_rows, img_cols)))
model.add(Flatten())
model.add(Dense(1024))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(nb_classes))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
optimizer='adadelta',
metrics=['accuracy'])
```

I tried to change the loss to hinge

```
model = Sequential()
model.add(Convolution2D(nb_filters, nb_conv, nb_conv,
                            border_mode='valid',
                            input_shape=(1, img_rows, img_cols)))
model.add(Flatten())
model.add(Dense(1024))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(nb_classes))
model.add(Activation('linear'))

model.compile(loss='hinge',
              optimizer='adadelta',
              metrics=['accuracy'])
```

but the accuracy of training data doesn't change in each iteration

**Comment by erlendd** [comment]
I'm not sure linear in the correct activation on your output: I think it should be tanh so that you get labels [-1,+1].

**Comment by alyato** [comment]
@mundher Do you solve your problem? When i want to use the svm, i always call  the scikit-learn.svm.How do you implement it ? Thanks.

**Comment by mundher** [comment]
@alyato I didn't solve the problem. it looks like the hinge loss works only for one class output

**Comment by fish128** [comment]
I tried a few different activations at the last layer, and found that the softmax activation works best with hinge loss. Can anyone explain this?

**Comment by huanglianghua** [comment]
You need to regularize the weight. 

The hinge loss with regularization term forms the complete SVM loss function. Try:

```python
from keras.regularizers import l2

model = Sequential()
model.add(Convolution2D(nb_filters, nb_conv, nb_conv,
                            border_mode='valid',
                            input_shape=(1, img_rows, img_cols)))
model.add(Flatten())
model.add(Dense(1024))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(nb_classes), W_regularizer=l2(0.01))
model.add(Activation('linear'))

model.compile(loss='hinge',
              optimizer='adadelta',
              metrics=['accuracy'])
```

instead.

**Comment by McLawrence** [comment]
in the code the hinge loss is defined as:
`K.mean(K.maximum(1. - y_true * y_pred, 0.), axis=-1)`

However, as I know it, the loss for an SVM should be this:

![screen shot 2017-04-29 at 16 54 25](https://cloud.githubusercontent.com/assets/25885486/25556452/101f3f24-2cfd-11e7-97e4-491cc9f5a2b0.png)

where s_i ist the score of the i-th output unit. As I understand it the hingeloss ignores the additive s_j term and only uses the s_{y_i}-term in the formula. Where is my error?

**Comment by aniket03** [comment]
@McLawrence  the hinge loss implemented in keras is for a specific case of binary classification [A vs ~A]. If this is to be used labels must be in the format of {-1, 1}

**Comment by gledsonmelotti** [comment]
How are you? I am trying to implement an SVM within the cost function using from sklearn.svm import SVC from sklearn.model_selection import GridSearchCV However, I always get errors. I believe the problem is to convert the y_true and y_pred tensor into a numpy array to be used in sklearn.svm. Then I need to convert the predicted results to tensor to be used in the cost function of keras (categorical_hinge).

Can anybody help me?

model_input = Input(shape = (img_width, img_height, channel_axis))
x = Convolution2D_bn(model_input, 32, 3, 3, strides=(2, 2), padding='valid')
x = Convolution2D_bn(x, 32, 3, 3, padding='valid')
x = Convolution2D_bn(x, 64, 3, 3)
x = MaxPooling2D((3, 3), strides=(2, 2))(x)

x = Convolution2D_bn(x, 80, 1, 1, padding='valid')
x = Convolution2D_bn(x, 192, 3, 3, padding='valid')
more model
# Classification block
x = GlobalAveragePooling2D()(x)
x = Dense(4096, kernel_regularizer=l2(1e-4), name='Dense_1')(x)
x = Activation('relu', name='relu1')(x)
x = Dropout(DROPOUT)(x)
x = Dense(4096, kernel_regularizer=l2(1e-4), name='Dense_2')(x)
x = Activation('relu', name='relu2')(x)
model_output = Dropout(DROPOUT)(x)
model = Model(model_input, model_output)
model.summary()

import tensorflow as tf
from keras import backend as K
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import StandardScaler
from keras.losses import categorical_hinge
def custom_loss_value(y_true, y_pred):
    X = K.eval(y_pred)
    print(X)
    Y = np.ravel(K.eval(y_true))
    Predict = []
    Prob = []
    scaler = StandardScaler()
    X = scaler.fit_transform(X)
    param_grid = {'C': [0.1, 1, 8, 10], 'gamma': [0.001, 0.01, 0.1, 1]}
    SVM = GridSearchCV(SVC(kernel='rbf',probability=True), cv=3, param_grid=param_grid, scoring='auc', verbose=1)
    SVM.fit(X, Y)
    Final_Model = SVM.best_estimator_
    Predict = Final_Model.predict(X)
    Prob = Final_Model.predict_proba(X)
    return categorical_hinge(tf.convert_to_tensor(Y, dtype=tf.float32), tf.convert_to_tensor(Predict, dtype=tf.float32))

sgd = SGD(lr=0.001, decay=1e-6, momentum=0.9, nesterov=True) 
model.compile(loss=custom_loss_value, optimizer=sgd, metrics=['accuracy'])

**Comment by bit-scientist** [comment]
@huanglianghua , @aniket03 Is it really enough to change the loss in the compile to `hinge` and regularize the last Dense layer with any ( like this,  `kernel_regularizer=regularizers.l2(0.01)`) penalty? I can't seem to understand that? Could you cite some links, posts? Thanks

**Comment by statcom** [comment]
> @huanglianghua , @aniket03 Is it really enough to change the loss in the compile to `hinge` and regularize the last Dense layer with any ( like this, `kernel_regularizer=regularizers.l2(0.01)`) penalty? I can't seem to understand that? Could you cite some links, posts? Thanks

Here are a couple of good references::
https://cs231n.github.io/linear-classify/
https://github.com/nfmcclure/tensorflow_cookbook#ch-4-support-vector-machines

I tested the following code with real data without any problem. Of course, the results will be different from the ones from real SVM implementation (e.g., sklearn's SVM). An interesting thing is that this Keras implementation produced a lot better results with specific data sometimes. Of course, another benefit is that you can use GPU to train SVM.

    fine_model_st.add(Dense(nb_classes, kernel_regularizer=regularizers.l2(0.0001)))
    fine_model_st.add(Activation('linear'))
    fine_model_st.compile(loss='squared_hinge',
                          optimizer='adadelta', metrics=['accuracy'])

**Comment by momja** [comment]
How does this align with the use of RandomFourierFeatures for SVM approximation found [here](https://keras.io/examples/keras_recipes/quasi_svm/)? Is RandomFourierFeatures a better, more modern approach?

**Comment by Apidcloud** [comment]
The suggested code from @huanglianghua and @statcom seem to work, but I wonder if there is a way of outputting probabilities instead. I tried to use **softmax**, but the model doesn't improve at all (with 3 classes it gets stuck at 33%; with 2 classes gets stuck at 50%). My goal is to 'exactly' replicate an SVM model, so that I can convert it to .**onnx** and reuse it.

The only settings I got it to work are the following:

```
  # create model
  model = Sequential()
  # note sure whether the number of filters for the first layer needs to match the input shape or not
  model.add(Dense(30, input_shape=(30,), activation='relu', kernel_initializer='he_uniform'))
  model.add(Dense(count_classes, kernel_regularizer=regularizers.l2(0.1)))
  #model.add(Activation('softmax')) # linear by default; softmax doesn't seem to work. Any ideas?
  model.compile(loss=keras.losses.CategoricalHinge(), optimizer=keras.optimizers.Adam(lr=1e-3), metrics=['accuracy'])
```

Any ideas on how to get a probability at the end? Using **argmax** seems to work just fine, but how do I interpret the output? I mean, I would like to set some sort of threshold to decide whether a prediction is good or not. Let's say I got 3 classes, but I input something else to the model to predict. I will still get a prediction, even if it's wrong. A probability would let me avoid this by setting some sort of threshold. Any ideas on how to approach this?

**Comment by erlendd** [comment]
Gettng probabilities out of SVM is usually done by adding a logistic regression after the linear output of the base SVM model. Use a `K.stop_gradient` to prevent the logistic layer from affecting the weights in the base model.

**Comment by Apidcloud** [comment]
Thanks, @erlendd 

I tried the following approach:
```
  input = Input(shape=(30,))
  dense = Dense(30, activation='relu', kernel_initializer='he_uniform', name='mul')
  x = dense(input)
  x = Dense(count_classes, kernel_regularizer=regularizers.l2(0.1), name='regul')(x)
  stop_grad = Lambda(lambda x: K.stop_gradient(x))(x)
  x = Activation('linear')(stop_grad)
  
  output = Dense(count_classes, activation="softmax", name='out')(x)

  model = Model(inputs=input, outputs=output)
  model.compile(loss=keras.losses.CategoricalHinge(), optimizer=keras.optimizers.Adam(lr=1e-3), metrics=['accuracy'])
```

But it still gets stuck at 33% for some reason.

![image](https://user-images.githubusercontent.com/3720818/118481851-dae69580-b70b-11eb-9efd-a4772b104fcd.png)
